### DISCUSSION TITLE
How to get a string from an inline array in a struct?

### DISCUSSION THREAD

**Discussion by Lightczx** [microsoft/CsWin32#635] [Q&A] [answered]
How can I get the MODULEENTRY32.szModule 's `string` representation ?
It's currently defined as `__winmdroot_Foundation_CHAR_256` struct type.

**Comment by AArnott** [comment, marked as answer]
If this were using a UTF-16 character array, there would be a ToString() method on the struct that would produce what you want.
But as this is a `CHAR` inline array, you have to do it yourself. I filed #636 to improve this.

This is how you can do it. It produces the string "abc":

```cs
using System.Text;
using Windows.Win32;
using Windows.Win32.Foundation;
using Windows.Win32.System.Diagnostics.ToolHelp;

MODULEENTRY32 m = default;
m.szModule._0 = (CHAR)(byte)'a';
m.szModule._1 = (CHAR)(byte)'b';
m.szModule._2 = (CHAR)(byte)'c';

unsafe
{
    byte* pszModule = (byte*)&m.szModule;
    string szModule = Encoding.UTF8.GetString(pszModule, GetLength(pszModule));
}

static unsafe int GetLength(byte* pszStr)
{
    int len = 0;
    while (*pszStr++ != 0)
    {
        len += 1;
    }

    return len;
}
```

**Reply by Lightczx** [reply]
Thanks, It works properly!

**Comment by Lightczx** [comment]
C# 11 introduced a new feature called `UTF-8 string literals`.
As it now we can use utf8 string easily, I just wrote a method to get the "string"

``` c#
public static unsafe ReadOnlySpan<byte> AsNullTerminatedReadOnlySpan(this in __CHAR_256 char256)
{
    fixed (CHAR* pszChar = &char256._0)
    {
        return MemoryMarshal.CreateReadOnlySpanFromNullTerminated((byte*)pszChar);
    }
}
```

**Reply by jnm2** [reply]
Unless `__CHAR_256` is a `ref` struct and therefore limited to the stack, this code is asking for trouble. As soon as the `fixed` block ends, the address of `char256._0` could change during GC but the span returned by this method would not be updated. (This is a GC hole.)

It's very different if you're making a copy within the `fixed` block, for instance to return `string`.

**Reply by Lightczx** [reply]
Yes, it's dangerous. So I use it like the `AsReadOnlySpan` method's replacement in the struct.

**Reply by AArnott** [reply]
Why not use `AsReadOnlySpan` then? That method being in the struct (and defining certain attributes or the new modifier) allows C# to help the caller stay safe, but it looks like your extension method will circumvent C#'s protections.

**Reply by AArnott** [reply]
If it's just that you want the span sized given null-termination, I imagine we could add that for CHAR-based inline array structs.

**Reply by jnm2** [reply]
I think it would be possible to create a span to the actual ref instead of to a pointer, where there would be no GC hole, but I don't remember how. I do know that the Span API has new `ref` constructors for a single item but no count IIRC.

**Reply by AArnott** [reply]
Yes, it should be possible. And we already do. That's why I wonder why this extension method need exist at all. If it's just so the span length is set till the null terminator instead of the full fixed array's length, that's something we can make easy for CHAR inline array structs just as we do for `char` ones.

**Reply by Lightczx** [reply]
It‘s a semantical issue:
I don't want lead to unnecessary memory allocation then:
If I use AsReadOnlySpan, it's a bit hard to compare with certain `ReadOnlySpan<byte>` like `"modulePath.dll"u8`
I have to cast the `ReadOnlySpan<CHAR>` to `ReadonlySpan<byte>` using `MemoryMarshal.AsBytes` then use `MemoryMarshal.IndexOf` to compare between them.
If there is API return null terminated readonly span of byte, I can just directly use `MemoryMarshal.SequenceEqual` and the code reads more smooth though.

After all, it's just something semantical.
